### PR TITLE
integrate `tags` into PDS and App View

### PR DESCRIPTION
**Pull request by estrattonbailey**
This PR handles creating posts with tags, validation of those tags, indexing into our db, and returning on the post thread view.

### CODE REVIEW

File: packages/pds/src/api/com/atproto/repo/createRecord.ts [hunk 1]
```diff
@@ -44,6 +46,16 @@ export default function (server: Server, ctx: AppContext) {
           'Unvalidated writes are not yet supported.',
         )
       }
+      if (AppBskyFeedPost.isRecord(record)) {
+        if (record.tags) {
+          for (const tag of record.tags) {
+            if (!isValidHashtag(tag)) {
+              throw new InvalidRequestError(`Invalid hashtag: ${tag}`)
+            }
```
Comment (estrattonbailey) [inline packages/pds/src/api/com/atproto/repo/createRecord.ts:54]: Checks for slurs here

File: packages/pds/src/handle/index.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import * as ident from '@atproto/syntax'
 import { InvalidRequestError } from '@atproto/xrpc-server'
 import { reservedSubdomains } from './reserved'
-import { hasExplicitSlur } from './explicit-slurs'
+import { hasExplicitSlur } from '../util/explicit-slurs'
```
Comment (estrattonbailey) [inline packages/pds/src/handle/index.ts:4]: Moved this to a more generic place, added `utils/validation.ts` also

File: packages/bsky/src/api/app/bsky/feed/getAuthorFeed.ts [hunk 3]
```diff
@@ -104,6 +105,17 @@ export const skeleton = async (
...
 
+  if (tagsString) {
+    const tags = tagsString.split(',')
+    feedItemsQb = feedItemsQb.where(
+      'post.tags',
+      '@>',
+      sql`CAST(${JSON.stringify(tags)} as jsonb)`,
+    ) // AND logic
+    // @ts-ignore TODO
+    // feedItemsQb = feedItemsQb.where('post.tags', '?|', sql`CAST(${JSON.stringify(tags)} as jsonb)`) // OR logic, doesn't work
```
Comment (estrattonbailey) [inline packages/bsky/src/api/app/bsky/feed/getAuthorFeed.ts:116]: So using the `@>` JSONB operator works fine, but applies AND logic.

To apply OR logic, we need like `where <jsonb> ?| array['tag', 'tag2']`, but I'm struggling to produce that with Kysely.

Halp 😅

File: packages/pds/src/api/com/atproto/repo/createRecord.ts (hunk 1 above)
Comment (dholms) [reply packages/pds/src/api/com/atproto/repo/createRecord.ts:54]: we do this check over in `src/repo/prepare.ts` - would be good to keep it all in one spot 👌 

also would be good to mirror the check over in the auto-moderator code in the appview

File: packages/pds/src/repo/prepare.ts [hunk 4]
```diff
@@ -299,6 +299,8 @@ function assertNoExplicitSlurs(rkey: string, record: RepoRecord) {
   } else if (isFeedGenerator(record)) {
     toCheck += ' ' + rkey
     toCheck += ' ' + record.displayName
+  } else if (isPost(record)) {
+    toCheck += record.tags?.join(' ')
```
Comment (estrattonbailey) [inline packages/pds/src/repo/prepare.ts:303]: Moved validation 👍

File: packages/pds/src/api/com/atproto/repo/createRecord.ts (hunk 1 above)
Comment (estrattonbailey) [reply packages/pds/src/api/com/atproto/repo/createRecord.ts]: Done!

File: packages/bsky/src/api/app/bsky/feed/getAuthorFeed.ts [hunk 5]
```diff
@@ -54,7 +55,7 @@ export const skeleton = async (
   params: Params,
   ctx: Context,
 ): Promise<SkeletonState> => {
-  const { cursor, limit, actor, filter, viewer } = params
+  const { cursor, limit, actor, filter, viewer, tags: tagsString } = params
```
[the remark this answers is not in the document]
Comment (estrattonbailey) [reply packages/bsky/src/api/app/bsky/feed/getAuthorFeed.ts]: Went ahead and removed this, easy enough to add back if we want to 👍

**Comment by estrattonbailey** [comment]
Gonna just merge this into the base branch for easier reviewing 👍

**Pull request merged**